### DISCUSSION TITLE
How to migrate observe() to lit

### DISCUSSION THREAD

**Discussion by paladox** [lit/lit#2319] [Q&A] [answered]
Hi, so my question is how do i migrate something like

```
  @observe('rule.value.*')
  _handleValueChange() {
    if (!this._originalRuleValues || !this.rule?.value) {
      return;
    }
    this.rule.value.modified = true;
    // Allows overall access page to know a change has been made.
    fireEvent(this, 'access-modified');
  }
```
to lit?

I understand that using willUpdate/updated doesn't work with objects and also doesn't work if you want to do things like rule.value.*.

**Comment by kevinpschaaf** [comment, marked as answer]
This is similar in spirit to the discussion in #2304 so I'll post similar answers:

Lit has a simple dirty-checking model, where (by default) a strict equality check (`===`) is used to trigger the update cycle when a `@property`/`@state` field is set. In other words, the value of the field _must_ change to trigger an update, and the previous value is provided in the `changedProperties` passed to `willUpdate`/`updated`. 

This system is much simpler than Polymer's and well-suited to "immutable data" patterns where root objects are created anew whenever a sub-property is changed (e.g. `this.value = {...this.value, someNewProp: 'foo'}`) rather than mutating the sub-property directly. So this is the first step: ensuring that immutable patterns are used rather than deep modifications of objects, to ensure Lit sees the change to the sub-property by means of the root property changing and triggers an update.

Second, because Lit doesn't store previous sub-property values or have a way of triggering an update based on a sub-property value changing, it doesn't have a way to give you the same information that Polymer did on its own. So in order to know that `rule.value.*` changed (as opposed to e.g. `rule.foo` or `rule.bar.baz`, you would need to allocate your own storage to hold the previous `rule.value` and do the dirty checking yourself. Something like:

```js
prevRuleValue = undefined;

willUpdate(changedProperties) {
  if (this.rule?.value !== this.prevRuleValue) {
    this.prevRuleValue = this.rule?.value;
    this._handleValueChange();
  }
}
```
